kong/resty/dns/client.lua

```diff
@@ -839,7 +813,37 @@ local function syncQuery(qname, r_opts, try_list)
   local key = qname..":"..r_opts.qtype
   local item = queue[key]
 
-  -- if nothing is in progress, we start a new async query
+  local is_yieldable = in_yieldable_phase()
+
+  -- If nothing is in progress and in the yieldable phase,
+  -- we start a new sync query
+
+  if not item and is_yieldable then
+    item = {
+      key = key,
+      semaphore = semaphore(),
+      qname = qname,
+      r_opts = cycle_aware_deep_copy(r_opts),
+      try_list = try_list,
+    }
+    queue[key] = item
+
+    item.result, item.err = individualQuery(qname, item.r_opts, try_list)
+
+    -- query done, but by now many others might be waiting for our result.
+    -- 1) stop new ones from adding to our lock/semaphore
+    queue[key] = nil
+    -- 2) release all waiting threads
+    item.semaphore:post(math_max(item.semaphore:count() * -1, 1))
+    item.semaphore = nil
+    ngx.sleep(0)
```

Comment:
Should we use `require("kong.tools.yield").yield()`?

Reply:
yeah, good catch, using `yield` api is better, will update later

Reply:
fixed